### PR TITLE
Remove ability to plug-in TransportService

### PR DESCRIPTION
**Pull request by s1monw**
TransportService is such a central part of the core server, replacing
it's implementation is risky and can cause serious issues. This change removes the ability to
plug in TransportService but allows registering a `TransportInterceptor` that enables
plugins to intercept requests on both the sender and the receiver ends. This is a commonly used
and overwritten functionality but encapsulates the custom code in a contained manner.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/common/network/NetworkModule.java [hunk 1]
```diff
@@ -181,4 +175,39 @@ private void registerBuiltinAllocationCommands() {
     public boolean canRegisterHttpExtensions() {
         return transportClient == false;
     }
+
+    /**
+     * Registers a new {@link org.elasticsearch.transport.TransportService.TransportInterceptor}
+     */
+    public void addTransportInterceptor(TransportService.TransportInterceptor interceptor) {
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/common/network/NetworkModule.java]: Is this just temporary, and can/should be replaced with a pull based interface in a followup?

File: core/src/main/java/org/elasticsearch/transport/TransportService.java [hunk 2]
```diff
@@ -1065,6 +1069,41 @@ public long getRequestId() {
...
+        }
 
+        /**
+         * This is called up-front providing the actual low level {@link AsyncSender} that performs the low level send request.
+         * The returned sender is used to send all requests that come in via
+         * {@link #sendRequest(DiscoveryNode, String, TransportRequest, TransportResponseHandler)} or
+         * {@link #sendRequest(DiscoveryNode, String, TransportRequest, TransportRequestOptions, TransportResponseHandler)}.
+         * This allows plugins to perform actions on each send request including modifying the request context etc.
+         */
+        default AsyncSender asyncSender(AsyncSender sender) {
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/transport/TransportService.java]: nit: can this be named in parallel, ie interceptSender?

**Review by rjernst — APPROVED** [review]
LGTM. Perhaps we could/should move TransportInterceptor out from an inner class? My thought is just that plugins knowing about the TransportService is not necessary (TransportService simply uses the new interface).

File: core/src/main/java/org/elasticsearch/transport/TransportService.java (hunk 2 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/transport/TransportService.java]: sure

File: core/src/main/java/org/elasticsearch/common/network/NetworkModule.java (hunk 1 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/common/network/NetworkModule.java]: yeah I didn't wanna do that at the same time

**Comment by s1monw** [comment]
@rjernst I applied your comments

**Comment by rjernst** [comment]
Still LGTM

**Pull request merged**